esp-hal/src/dma/mod.rs

```diff
@@ -902,141 +953,262 @@ impl DescriptorChain {
             return Err(DmaError::UnsupportedMemoryRegion);
         }
 
-        if self.descriptors.len() < len.div_ceil(self.chunk_size) {
-            return Err(DmaError::OutOfDescriptors);
-        }
+        let max_chunk_size = self.max_chunk_size(len, circular)?;
 
-        if circular && len <= 3 {
-            return Err(DmaError::BufferTooSmall);
-        }
+        DescriptorSet::set_up_buffer_ptrs(
+            unsafe { core::slice::from_raw_parts_mut(data, len) },
+            self.descriptors,
+            max_chunk_size,
+            circular,
+        )?;
+        DescriptorSet::set_up_descriptors(
+            self.descriptors,
+            len,
+            max_chunk_size,
+            circular,
+            prepare_descriptor,
+        )?;
 
-        self.descriptors.fill(DmaDescriptor::EMPTY);
+        Ok(())
+    }
 
-        let max_chunk_size = if !circular || len > self.chunk_size * 2 {
-            self.chunk_size
-        } else {
+    fn max_chunk_size(&self, len: usize, circular: bool) -> Result<usize, DmaError> {
+        let max_chunk_size = if circular && len <= self.chunk_size * 2 {
+            if len <= 3 {
+                return Err(DmaError::BufferTooSmall);
+            }
             len / 3 + len % 3
+        } else {
+            self.chunk_size
         };
 
-        let mut processed = 0;
-        let mut descr = 0;
-        loop {
-            let chunk_size = usize::min(max_chunk_size, len - processed);
-            let last = processed + chunk_size >= len;
+        Ok(max_chunk_size)
+    }
+}
 
-            let next = if last {
-                if circular {
-                    addr_of_mut!(self.descriptors[0])
-                } else {
-                    core::ptr::null_mut()
-                }
-            } else {
-                addr_of_mut!(self.descriptors[descr + 1])
-            };
+/// Computes the number of descriptors required for a given buffer size with
+/// a given chunk size.
+pub const fn descriptor_count(buffer_size: usize, chunk_size: usize, is_circular: bool) -> usize {
+    if is_circular && buffer_size <= chunk_size * 2 {
+        return 3;
+    }
 
-            // buffer flags
-            let dw0 = &mut self.descriptors[descr];
+    if buffer_size < chunk_size {
+        // At least one descriptor is always required.
+        return 1;
+    }
 
-            dw0.set_suc_eof(false);
-            dw0.set_owner(Owner::Dma);
-            dw0.set_size(chunk_size); // align to 32 bits?
-            dw0.set_length(0); // hardware will fill in the received number of bytes
+    buffer_size.div_ceil(chunk_size)
+}
 
-            // pointer to current data
-            dw0.buffer = unsafe { data.add(processed) };
+/// Compute max chunk size based on block size.
+const fn chunk_size(block_size: Option<DmaBufBlkSize>) -> usize {
```

Comment:
To match the comment
```suggestion
const fn max_chunk_size(block_size: Option<DmaBufBlkSize>) -> usize {
```